### ISSUE TITLE
save() might result in an error if the object is not dirty

### ISSUE THREAD

**Issue by lhazlewood** [stormpath/stormpath-sdk-java#27]
The following code with 0.9.1 results in an error:

```
Account newAccount = createAccount(adminUser); 
String href = createAccountInDirectory(accountDirectory, newAccount);

Group realmAdminGroup = getClient().getDataStore()
    .getResource(accountDirectory.getAdminGroupHref(), Group.class);

newAccount.addGroup(realmAdminGroup); 
newAccount.save();
```

I suspect that save() is called, and the account resource itself is not dirty, and the error occurs because of that.

**Comment by markwoon** [comment]
I would really like to be able to call() save without having to worry about whether the object is dirty.  That's just one extra thing to worry about.

Unless you can provide an isDirty() check?

At the very least, I think the error message should be changed to specifically state that there were no changes, as opposed to "Account data cannot be null, empty, or blank".   Too bad we can't use status 304 to indicate something like this.  Or an alternate Stormpath code so that I can easily catch and ignore these errors.

**Comment by mrioan** [comment]
AbstractResource res = (AbstractResource) account;
if (res.isDirty()) {
   account.save();

}

**Comment by mrioan** [comment]
I agree that the error propagated to the developer is very generic. We will consider wrapping it with a more precise message.

In the meantime, I think this workaround can be helpful for you:

```java
AbstractResource resource = (AbstractResource) account;
if (resource.isDirty()) {
    account.save();
}
```

**Comment by lhazlewood** [comment]
That's a workaround, but not ideal: it requires a compile-time dependency on the Stormpath `impl` .jar, which is specifically designed *not* to allow compile-time dependencies to ensure backwards-incompatible changes do not impact SDK users.

That said, @mrioan's code example can be used as a temporary workaround, but we'll absolutely address this formally in the SDK API.  Since being 'dirty' is an implementation detail, my current belief is that such checks shouldn't be exposed and `save()` just return quietly.

**Comment by markwoon** [comment]
Sounds good.

Note that `CustomData.save()` also returns silently if nothing has changed.  So this behavior would be more consistent.

**Comment by markwoon** [comment]
BTW, if you make `isDirty()` part of the public API I can think of a few places it would come in handy for me...

**Comment by dogeared** [comment]
@lhazlewood @mrioan - Is this still relevant? `isDirty` still only exists in `AbstractResource`, which is in the `impl` package.
